content/blog/2022/2022-08-03-feature-flags-in-nodejs-react-client-with-launchdarkly.md

```diff
@@ -0,0 +1,320 @@
+
+As discussed above, we will fetch the *Client-Side ID* and add it in the code.
+
+{{% image alt="LaunchDarkly Keys" src="images/posts/nodejs-frontend-launchdarkly/LaunchDarkly_Keys.png" %}}
+
+Next we will update our `App.js` to fetch the feature flag variations from LaunchDarkly and display that button: 
+
+```javascript
+import React, { Component } from 'react';
+import styled from "styled-components";
+import './App.css';
+import * as LDClient from 'launchdarkly-js-client-sdk';
+
+const theme = {
+  blue: {
+    default: "#3f51b5",
+    hover: "#283593"
+  }
+};
+
+const Button = styled.button`
+    background-color: ${(props) => theme[props.theme].default};
+    color: white;
+    padding: 5px 15px;
+    border-radius: 5px;
+    outline: 0;
+    text-transform: uppercase;
+    margin: 10px 0px;
+    cursor: pointer;
+    box-shadow: 0px 2px 2px lightgray;
+    transition: ease background-color 250ms;
+    &:hover {
+      background-color: ${(props) => theme[props.theme].hover};
+    }
+    &:disabled {
+      cursor: default;
+      opacity: 0.7;
+    }
+  `;
+
+const clickMe = () => {
+  alert("A new shiny feature pops up!");
+};
+
+class App extends Component {
+  constructor() {
+    super()
+    this.state = {}
+  }
+  componentDidMount() {
+    const user = {
+      // UI based logged-in user
+      key: 'john_doe'
+    }
+    // SDK requires Client-side ID for UI call
+    this.ldclient = LDClient.initialize('62e*********************', user);
+    this.ldclient.on('ready', this.onLaunchDarklyUpdated.bind(this));
+    this.ldclient.on('change', this.onLaunchDarklyUpdated.bind(this));
+  }
+  onLaunchDarklyUpdated() {
+    this.setState({
+      featureFlags: {
+        showShinyNewFeature: this.ldclient.variation('show-shiny-new-feature', false)
+      }
+    })
+  }
+  render() {
+    if (!this.state.featureFlags) {
+      return <div className="App">Loading....</div>
+    }
+    return (
+      <div className="App">
+        <div>{ 
+            this.state.featureFlags.hideUser
+              ? <Button theme='blue' onClick={clickMe}>Shiny New Feature</Button>
+              : '' }
+        </div>
+    );
+  }
+}
+
+export default App;
+```
+
+To execute code only when the LaunchDarkly client is ready, we use the `ldclient.on()` function.
+
+With `this.ldclient.on('ready', ...)`, we subscribe to the `ready` event which will fire once the LaunchDarkly client has received the state of all feature flags from the server.
+
+With `this.ldclient.on('change', ...)`, we subscribe to the `change` event which will fire once the LaunchDarkly client has received the state of all feature flags that has got updated or changed in the LaunchDarkly UI.
+
+Next we can run the following command to start the UI:
+
+```bash
+npm start
+```
+
+When the user is not set in the LaunchDarkly UI, our React UI looks like below:
```

Comment:
The image shows the "shiny new button" ... "when the user is not set", it should show an empty screen, shouldn't it? Can you add a screenshot of both states?

Reply:
It would show blank screen like below. Should we add it? 

![image](https://user-images.githubusercontent.com/8446841/193452601-92b2c121-9bc1-4331-ac4f-37a9514bad63.png)